src/explorer/webview.ts

```diff
@@ -729,104 +738,105 @@ class ExplorerWebview {
-                    let file: string;
-                    if (fLocation.uri !== undefined) {
-                        // @ts-ignore external exist on the webview side
-                        file = fLocation.uri.external.replace("%3A", ":");
-                    }
+        }
+        const tableHeadEle: HTMLHeadElement = this.createElement("thead");
+        tableHeadEle.appendChild(headerRow);
+        tableEle.appendChild(tableHeadEle);
 
-                    const fRow = this.createElement("tr", {
-                        attributes: {
-                            "data-eCol": fLocation.range[1].character.toString(),
-                            "data-eLine": fLocation.range[1].line.toString(),
-                            "data-file": file,
-                            "data-group": stackIndex,
-                            "data-sCol": fLocation.range[0].character.toString(),
-                            "data-sLine": fLocation.range[0].line.toString(),
-                        },
-                        className: "listtablerow",
-                    });
-
-                    if (file !== undefined) {
-                        fRow.addEventListener("click", this.onSourceLinkClickedBind);
-                    }
+        const tableBodyEle: HTMLBodyElement = this.createElement("tbody");
+        for (let stackIndex: number = 0; stackIndex < stacks.stacks.length; stackIndex++) {
+            const stack: Stack = stacks.stacks[stackIndex];
+            const msgRow: HTMLTableRowElement = this.createElement("tr", {
+                attributes: { "data-group": stackIndex, "tabindex": "0" },
+                className: `listtablegroup ${ToggleState.expanded}`,
+            });
+            msgRow.appendChild(this.createElement("th", {
+                attributes: { colspan: `${columnCount}` },
+                text: stack.message.text,
+            }));
+            msgRow.addEventListener("click", this.onToggleStackGroup.bind(this));
+            tableBodyEle.appendChild(msgRow);
+
+            const tdTag: string = "td";
+            for (const frame of stack.frames) {
+                const fLocation: Location = frame.location;
+                let file: string | undefined;
+                if (fLocation.uri) {
+                    // @ts-ignore external exist on the webview side
+                    file = fLocation.uri.external.replace("%3A", ":");
+                }
 
-                    const fLine = fLocation.range[0].line.toString();
-                    const fParameters = frame.parameters.toString();
-                    let fMsg = "";
-                    if (frame.message !== undefined && frame.message.text !== undefined) {
-                        fMsg = frame.message.text;
-                    }
+                const fRow: HTMLTableRowElement = this.createElement("tr", {
+                    attributes: {
+                        "data-eCol": fLocation.range !== undefined ? fLocation.range.end.character.toString() : 0,
```

Comment:
Another opportunity for reduction:`fLocation.range?.end.character.toString() ?? 0`.

Reply:
Actually, after ensuring that a range is always created (even if it is the first line), the undefined check is no longer needed.

---
In reply to: [397704938](https://github.com/microsoft/sarif-vscode-extension/pull/211#discussion_r397704938) [](ancestors = 397704938)